### PR TITLE
Port documentation to CMake [10864]

### PR DESCRIPTION
**Pull request by EduPonz**
This PR ports the documentation over to CMake:
* Makefile no longer exists, so its commands are not valid anymore
* Build is performed with either `cmake --build` or `colcon build`. This PR adds build options:
    * `BUILD_DOCUMENTATION` -> Whether to build the sphinx docs. Defaults ON
    * `COMPILE_TESTS` -> Whether to build the tests. Defaults ON
* With `colcon`/`cmake`, the documentation is built using Sphinx's HTML builder. Users can still use other builders using Sphinx directly.
* Tests are moved to CTest so they can be run with `ctest` or `colon test`
* Documentation is installed
* README.md has been updated to reflect the new building methods

Merging this PR entails that:
* Current CI jobs will need to be updated.
* It offers the possibility to test the documentation with every Fast DDS PR, since now it can be built and test with `colcon`. Doing so will prevent merging Fast DDS PR with spelling errors in the doxygen comments, which right now breaks the documentation tests every now and then.
* External contributions are made easier, since with this PR, everyone can build and run the documentation tests locally, whereas before there was a dependency with eProsima's intranet.

### CODE REVIEW

File: code/grep.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
+# Copyright 2016 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```
Comment (JLBuenoLopez-eProsima) [inline code/grep.py:1]:
```suggestion
# Copyright 2021 Proyectos y Sistemas de Mantenimiento SL (eProsima).
```

File: code/CMakeLists.txt [hunk 2]
```diff
@@ -12,243 +12,153 @@
...
+# Check docs style
+find_program(DOC8 docs.doc8 DOC "path to doc8 executable")
+message(STATUS "doc8 found: ${DOC8}")
+if(DOC8)
+    add_test(NAME doc8-test COMMAND ${DOC8} --max-line-length 120 ${CMAKE_SOURCE_DIR}/docs)
+else()
+    message(FATAL_ERROR "Could not find doc8")
 endif()
+
+# Check that all cpp snippets are on the CPP files so they are compiled
```
Comment (JLBuenoLopez-eProsima) [inline code/CMakeLists.txt:142]:
```suggestion
# Check that all cpp snippets are on the CPP files and not in the RST files so they are compiled
```

File: code/CMakeLists.txt [hunk 3]
```diff
@@ -12,243 +12,153 @@
...
+add_test(NAME docs.code_block_check.cpp
+    COMMAND
+    ${Python_EXECUTABLE} ${CMAKE_CURRENT_BINARY_DIR}/grep.py
+    --directory ${CMAKE_SOURCE_DIR}/docs
+    --regex "code-block::.*c[+p][+p]")
+
+# Check that all XML snippets are on the XML files so they are tested
```
Comment (JLBuenoLopez-eProsima) [inline code/CMakeLists.txt:149]:
```suggestion
# Check that all XML snippets are on the XML files and not in the RST files so they are tested
```

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -32,189 +32,243 @@ You can find all the library's source code on our [GitHub repository](https://gi
...
-    cd fastdds-docs-venv/lib/<python-version>/site-packages
-    curl https://patch-diff.githubusercontent.com/raw/sphinx-doc/sphinx/pull/7851.diff | git apply
-    cd -
+    mkdir -p <path_to_ws>/fastdds-docs_ws/src
+    cd <path_to_ws>/fastdds-docs_ws
+    wget https://raw.githubusercontent.com/eProsima/Fast-DDS-docs/master/colcon.meta
+    wget https://raw.githubusercontent.com/eProsima/Fast-DDS/master/fastrtps.repos
+    vcs import src < fastrtps.repos
+    cd src
+    git clone https://github.com/eProsima/Fast-RTPS-docs fastdds-docs
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply README.md:128]: I'm also inclined to that option. However, I'd wait until this PR is merged to do so, cause right now it does not make much sense. After this is merge, we can patch both repos

File: README.md [hunk 5]
```diff
@@ -32,189 +32,243 @@ You can find all the library's source code on our [GitHub repository](https://gi
...
-source fastdds-docs-venv/bin/activate
-FASTDDS_BRANCH=<branch> make <output_format>
+source <path_to_venv>/fastdds-docs-venv/bin/activate
+cd <path_to_repo>/fastdds-docs
+rm -rf build
+READTHEDOCS=True FASTDDS_BRANCH=<branch> sphinx-build \
+    -b html \
+    -D breathe_projects.FastDDS=<abs_path_to_docs_repo>/fastdds-docs/build/doxygen/xml \
+    -d <abs_path_to_docs_repo>/fastdds-docs/build/doctrees \
+    docs <abs_path_to_docs_repo>/fastdds-docs/build/html
```
[the remark this answers is not in the document]
Comment (EduPonz) [reply README.md:200]: I'm not able to reproduce this

**Comment by richiprosima** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/837/)

File: CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+        PROPERTY STRINGS "${FASTDDS_DOCS_ALLOWED_BUILDERS};${TWO_STEP_BUILDERS}")
+
+    # Check that specified value is allowed
+    message(STATUS "Selected builder: ${FASTDDS_DOCS_BUILDER}")
+    if((NOT FASTDDS_DOCS_BUILDER IN_LIST FASTDDS_DOCS_ALLOWED_BUILDERS) AND
+            (NOT FASTDDS_DOCS_BUILDER IN_LIST TWO_STEP_BUILDERS))
+        message(FATAL_ERROR,
+            "Wrong configuration of FASTDDS_DOCS_BUILDER.
+            Allowed values: ${FASTDDS_DOCS_ALLOWED_BUILDERS};${TWO_STEP_BUILDERS}")
+    endif()
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply CMakeLists.txt]: Finally, only `html` builder is going to be supported.

File: CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+option(BUILD_DOCUMENTATION "Generate documentation" ON)
+if (BUILD_DOCUMENTATION)
+    # Users can specify the sphinx builder with -DFASTDDS_DOCS_BUILDER. This option takes one value
+    # of the lists FASTDDS_DOCS_ALLOWED_BUILDERS or TWO_STEP_BUILDERS. This second list is for
+    # targets that require two steps in order to build the documentation. This way, the CMake poject
+    # can cover all the options that the Makefile had. FASTDDS_DOCS_BUILDER defaults to html
+    set(FASTDDS_DOCS_ALLOWED_BUILDERS
+        html dirhtml singlehtml pickle json htmlhelp
+        qthelp applehelp devhelp epub epub3 latex
+        text man texinfo gettext changes xml pseudoxml pdf)
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply CMakeLists.txt]: Only `html` builder is going to be supported.

File: Makefile [hunk 8]
```diff
@@ -1,261 +0,0 @@
...
-dummy:
-	$(SPHINXBUILD) -b dummy $(ALLSPHINXOPTS) $(BUILDDIR)/dummy
-	@echo
-	@echo "Build finished. Dummy builder generates no files."
-
-.PHONY: pdf
-pdf:
-	$(SPHINXBUILD) -d pdf $(ALLSPHINXOPTS) $(BUILDDIR)/pdf
-	@echo
-	@echo "Build finished. The PDF Files are in $(BUILDDIR)/pdf"
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply Makefile:261]: Only `html` builder is going to be supported.

File: README.md [hunk 9]
```diff
@@ -32,189 +32,243 @@ You can find all the library's source code on our [GitHub repository](https://gi
...
+| pickle     | Generate pickle files                                      |
+| json       | Generate JSON files                                        |
+| htmlhelp   | Generate HTML files and a HTML help project                |
+| qthelp     | Generate HTML files and a qthelp project                   |
+| applehelp  | Generate an Apple Help Book                                |
+| devhelp    | Generate HTML files and a Devhelp project                  |
+| epub       | Generate an epub                                           |
+| epub3      | Generate an epub3                                          |
+| latex      | Generate LaTeX files                                      |
+| latexpdf   | Generate LaTeX files and run them through pdflatex         |
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply README.md]: Only `html` builder is going to be supported.

File: README.md [hunk 10]
```diff
@@ -32,189 +32,243 @@ You can find all the library's source code on our [GitHub repository](https://gi
...
-1. [Running documentation tests](#running-documentation-tests)
-1. [Simulating Read the Docs](#simulating-read-the-docs)
+1. [Project structure](#project-structure)
 1. [Contributing](#contributing)
 
-## Project structure
+## Installation Guide
 
-The project is structured as follows:
+The following guide has been developed and tested in Ubuntu 20.04.
```
[the remark this answers is not in the document]
Comment (JLBuenoLopez-eProsima) [reply README.md:41]: I have not been able to build on Windows. On one hand, CMake is not able to find OpenSSL which is required. I have followed this [instructions](https://fast-dds.docs.eprosima.com/en/latest/installation/sources/sources_windows.html#openssl) but they are not working for me...

After disabling OpenSSL in the CMakeLists files, the following error appears:

```
CUSTOMBUILD : Configuration error : 
   There is a programmable error in your configuration file:
   OSError: [WinError 126] The specified module could not be found
```

Consequently, I agree to keep the statement that this has only been developed and tested on Ubuntu.

**Review by JLBuenoLopez-eProsima — COMMENTED** [review]
Just a couple of small things.

**Review by JLBuenoLopez-eProsima — DISMISSED** [review]
LGTM!

**Comment by richiprosima** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/840/)

**Comment by richiprosima** [comment]
Buildstatus:[BuildStatus](http://jenkins.eprosima.com:8080/job/fastdds_docs_manual/3/)

**Comment by richiprosima** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/841/)

**Comment by richiprosima** [comment]
Buildstatus:[BuildStatus](http://jenkins.eprosima.com:8080/job/fastdds_docs_manual/5/)

**Comment by richiprosima** [comment]
Build status: [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Docs%20Manual/842/)

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
Good catch!

**Comment by richiprosima** [comment]
Buildstatus:[BuildStatus](http://jenkins.eprosima.com:8080/job/fastdds_docs_manual/6/)

**Comment by richiprosima** [comment]
Buildstatus:[BuildStatus](http://jenkins.eprosima.com:8080/job/fastdds_docs_manual/7/)

**Pull request merged**